wasm/jit/engine.go

```diff
@@ -15,70 +15,327 @@ import (
-	stackBasePointer uint64
-	// Where we store the status code of JIT execution.
-	jitCallStatusCode jitCallStatusCode
-	// Set when statusCode == jitStatusCall{Function,BuiltInFunction}
-	// Indicating the function call index.
-	functionCallAddress wasm.FunctionAddress
-	// Set when statusCode == jitStatusCall{Function,BuiltInFunction}
-	// We use this value to continue the current function
-	// after calling the target function exits.
-	// Instructions after [base+continuationAddressOffset] must start with
-	// restoring reserved registeres.
-	continuationAddressOffset uintptr
-	// The current compiledFunction.globalSliceAddress
-	globalSliceAddress uintptr
-	// memorySliceLen stores the address of the first byte in the memory slice used by the currently executed function.
-	memorySliceAddress uintptr
-	// memorySliceLen stores the length of memory slice used by the currently executed function.
-	memorySliceLen uint64
-	// The current compiledFunction.tableSliceAddress
-	tableSliceAddress uintptr
-	// tableSliceLen stores the length of the unique table used by the currently executed function.
-	tableSliceLen uint64
-	// Function call frames in linked list
-	callFrameStack *callFrame
-	// callFrameNum tracks the current number of call frames.
-	// Note: this is not len(callFrameStack) because the stack is implemented as a linked list.
-	callFrameNum uint64
-
-	// The following fields are only used during compilation.
-
-	// Store the compiled functions.
-	compiledFunctions map[wasm.FunctionAddress]*compiledFunction
-}
+type (
+	engine struct {
+		// There contexts are read and written by JITed code.
+		// Note that we embed these structs so we can reduce the costs to access fields inside of them.
+		// Also, that eases the calculation of offsets to each field.
+		globalContext
+		moduleContext
+		valueStackContext
+		exitContext
+
+		// The following fields are NOT acceessed by JITed code directly.
+
+		// The actual Go-allocated stack for holding Wasm values.
+		// Note that we NEVER edit len or cap in JITed code so we won't get screwed when GC comes in.
+		valueStack []uint64
+
+		// Function call frames. The new one is set at callFrameStack[callFrameStackPoitner].
+		// The currently executed function call frame lives at callFrameStack[callFrameStackPoitner-1]
+		// and that is equivalent to  engine.callFrameTop().
+		callFrameStack []callFrame
+
+		// Store the compiled functions.
+		// The index means wasm.FunctionAddress, but we intentionally avoid using map
+		// as the underlying memory region is accessed by assembly directly by
+		// using compiledFunctionsFirstItemAddress.
+		compiledFunctions []*compiledFunction
+	}
+
+	// globalContext holds the data which stays same across multiple function calls.
+	globalContext struct {
+		// &engine.valueStack[0] as uintptr.
+		// Note: this is updated when growing the stack in builtinFunctionGrowValueStack.
+		valueStackFirstItemAddress uintptr
```

Comment:
here and everywhere as element is more often used than item

```suggestion
		valueStackElement0Address uintptr
```